wireguard/metrics.go

```diff
@@ -0,0 +1,258 @@
+
+	wireguardBytesSentFQName   = "wireguard_bytes_sent"
+	wireguardBytesSentHelpText = "wireguard interface total outgoing bytes from peer"
+
+	wireguardBytesRcvdFQName   = "wireguard_bytes_rcvd"
+	wireguardBytesRcvdHelpText = "total incoming bytes from peer"
+)
+
+func init() {
+	prometheus.MustRegister(
+		MustNewWireguardMetrics(),
+	)
+}
+
+type Metrics struct {
+	hostname           string
+	newWireguardClient func() (netlinkshim.Wireguard, error)
+	logCtx             *logrus.Entry
+
+	peerRx, peerTx map[wgtypes.Key]int64
+}
+
+func (collector *Metrics) Describe(chan<- *prometheus.Desc) {}
+
+func (collector *Metrics) Collect(m chan<- prometheus.Metric) {
+	collector.refreshStats(m)
+}
+
+func MustNewWireguardMetrics() *Metrics {
+	wg, err := NewWireguardMetrics()
+	if err != nil {
+		logrus.Panic(err)
+	}
+	return wg
+}
+
+func NewWireguardMetrics() (*Metrics, error) {
+	hostname, err := os.Hostname()
+	if err != nil {
+		logrus.WithError(err).Error("cannot register wireguard metrics stats")
+		return nil, err
+	}
+	return NewWireguardMetricsWithShims(hostname, netlinkshim.NewRealWireguard), nil
+}
+
+func NewWireguardMetricsWithShims(hostname string, newWireguardClient func() (netlinkshim.Wireguard, error)) *Metrics {
+	logrus.WithField("hostname", hostname).Debug("created wireguard collector for host")
+	return &Metrics{
+		hostname:           hostname,
+		newWireguardClient: newWireguardClient,
+		logCtx: logrus.WithFields(logrus.Fields{
+			"prometheus_collector": "wireguard",
+		}),
+
+		peerRx: map[wgtypes.Key]int64{},
+		peerTx: map[wgtypes.Key]int64{},
+	}
+}
+
+func (collector *Metrics) refreshStats(m chan<- prometheus.Metric) {
+	wgClient, err := collector.newWireguardClient()
+	if err != nil {
+		return
+	}
+	collector.collectDeviceMetrics(wgClient, m)
+	collector.collectPeerMetrics(wgClient, m)
+}
+
+func (collector *Metrics) collectDeviceMetrics(wgClient netlinkshim.Wireguard, m chan<- prometheus.Metric) {
+	collector.logCtx.Debug("collecting wg device metrics")
+
+	devices, err := wgClient.Devices()
+	if err != nil {
+		collector.logCtx.WithError(err).Error("error listing wireguard devices")
+		return
+	}
+	collector.logCtx.WithFields(logrus.Fields{
+		"count": len(devices),
+		"dev":   devices,
+	}).Debug("collect device metrics enumerated devices")
+
+	for _, device := range devices {
+		l := collector.defaultLabelValues(device.PublicKey.String())
+		for k, v := range deviceMetaLabelValues(device) {
+			l[k] = v
+		}
+		collector.logCtx.WithFields(logrus.Fields{
+			"dev":    device.Name,
+			"labels": l,
+		}).Debug("iterate device")
+
+		m <- prometheus.MustNewConstMetric(
+			prometheus.NewDesc(
+				wireguardMetaFQName, wireguardMetaHelpText, nil, l,
+			),
+			prometheus.GaugeValue,
+			1,
+		)
+	}
+}
+
+func (collector *Metrics) collectPeerMetrics(wgClient netlinkshim.Wireguard, m chan<- prometheus.Metric) {
+	collector.logCtx.Debug("collecting wg peer(s) metrics")
+
+	devices, err := wgClient.Devices()
```

Comment:
I think it's worth it for readability, but you're querying for devices here and also back in `collectDeviceMetrics`

Reply:
i moved both queries to refreshStats :)